### ISSUE TITLE
a error will occur in .smi-> .xyz->.smi

### ISSUE THREAD

**Issue by DreamMemory001** [openbabel/openbabel#2484]
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

openbabel==2.4.1

I want to convert .smi to .xyz, then i will achieve some task of prediciiton, so it is strict with the accuracy of 3D coordinate. When i use the openbabel convert .smi to .xyz, then convert .xyz to .smi. But i find the two smiles is differernt. the below is my code, I want to know  how to get the better accuracy in .xyz coordinate. Thank you!

.smi->.xyz
```python
## Use openbabel to convert SMILES to xyz.
mol = pybel.readstring("smi", smiles)
mol.localopt(forcefield="mmff94", steps=5000)
mol.make3D(forcefield='mmff94', steps=100)
mol.write("xyz", filename=out_name+'_pybel.xyz', overwrite=True)
```

.xyz->.smi

```python
def MolFormatConversion(input_file:str,output_file:str,input_format="xyz",output_format="smi"):
    molecules = readfile(input_format,input_file)
    output_file_writer = Outputfile(output_format,output_file)
    for i,molecule in enumerate(molecules):
        output_file_writer.write(molecule)
    output_file_writer.close()
    print('%d molecules converted'%(i+1))
```
This is the right smiles:
![image](https://user-images.githubusercontent.com/32425458/172286276-53de501c-a049-4c1c-b908-b742bb8c60c2.png)
This is convert smiles:
![image](https://user-images.githubusercontent.com/32425458/172286420-71022838-b981-42b9-9415-b471416e4e25.png)

**Comment by nbehrnd** [comment]
This is more a commentary to improve the presentation of your question, than as to provide a solution:
+ Update openbabel.  With version 2.4.1 you mention, you limit your work to what was implemented by 2016.  Though your code snippet to translate .smi into .xyz still is functional today, the present stable version is 3.1.1 ([reference](https://github.com/openbabel/openbabel/releases)).  Because it is possible that if you encounter a problem (or a bug in the underlying source code), this has been addressed since then.

+ You face a problem with the second example converting the .xyz into a .smi because the representation of the structure contains a pentavalent carbon.  Pick up the the two .xyz files (of both examples) you want to convert into SMILES, stash these files into a folder in common, and submit them as .zip archive by drag-and-drop.  By this complement to your question, you allow others to replicate your conversion (and perhaps the newer/contemporary version of openbabel trades this conversion better than version 2.4.1).

**Comment by fredrikw** [comment]
In addition, the original SMILES looks a bit strange with both the nitrogen and sulphur on the lower right of your image being without hydrogens.
And do you need the .xyz-format? There are other formats that handle chemical information a lot better, now OpenBabel has to guess the bonding based on the present atoms and with the strange radicals in the input that is difficult. Should you use e.g. .mol-format, you would get the same output molecule as you put in.
So, to summarize:
1. Update OpenBabel
2. Check your input SMILES
3. Reconsider using .xyz

**Comment by nbehrnd** [comment]
@fredrikw Good point to spot the missing H in the first structure _as if_ the structures might subject to keto-enol tautomerism.  In fact, submitting the SMILES assigned by ChemDraw to one (putative) structure of these pteridines

![pteridine](https://user-images.githubusercontent.com/19709084/172549678-aa88ec3b-262c-4b6f-af6e-b4b30dc46d58.png)

into [NIH's tautomerizer](https://cactus.nci.nih.gov/cgi-bin/tautomerize.tcl) a set of six isomers (compilation attached below).

@DreamMemory001 If there is no better format than .xyz accessible to you (better as in «retains bond orders _in addition_ to atoms' coordinates, like e.g., .mol, .mol2, .sdf), you might consider to check for plausibility of the results openbabel provides with a tool outside (working independently) of openbabel.  Because every algorithm requested to enrich the information stored in the records may err once on one, or an other side.  Though not leading to SMILES (but .mol files one can express again with SMILES), for «typical organic molecules»* Jan Jensen's [xyz2mol](https://github.com/jensengroup/xyz2mol) may assist here.  It was presented on [RDKit UGM 2020](https://github.com/rdkit/UGM_2020/blob/master/Presentations/JanJensen.pdf), his [video recording](https://www.youtube.com/watch?v=HD6IpXMVKeo) about it still is on youtube.

*) Known complications are salts, and difficulties with complexes including transition metals, etc.  So it is working best for neutral molecules consisting of (C, N, O, P, S; H; halogens).

[Predicted tautomers.pdf](https://github.com/openbabel/openbabel/files/8859045/Predicted.tautomers.pdf)

**Comment by DreamMemory001** [comment]
@nbehrnd @fredrikw 
Thanks for your reply.
* I have update the version of openbabel 3.1.1, but the result is not fit. In the document 3.0.0, I find the openbabel provide a new method base on paper "Fast, efficient fragment-based coordinate generation for Open Babel". But i can not find the how to use it.
*  "The nitrogen and sulphur on the lower right of my image being without hydrogens." In this question, i want to explain that because i need to genereate the molecules base on this scaffold. It is the scaffold. In my project, i predict the atomic number and the coordinate to identify a molecule, so i need .xyz file to save my result and the point where without hydrogens is the point that need to add other side chain. 
* I also use the rdkit (ETKDG), it is result is also not fit.
* Now, i use the dataset named QM9, the origin paper shows the idea of utilize smiles to get 3D coordinate. But it looks like diffcult to achieve it. And the all .xyz file in QM9 can convert .xyz to smiles rightly by openabel. So you have some other ideas in it? Thank you very much. Openbabel is a fantastic tool, i also hope it can become better and better. 

![image](https://user-images.githubusercontent.com/32425458/172571487-7ec24cd3-8e74-4f95-8621-568d14c13ae8.png)

* @nbehrnd Ok , i will try it by Jan Jensen's [xyz2mol](https://github.com/jensengroup/xyz2mol), Thank you.

**Comment by fredrikw** [comment]
I think you should rethink your approach a bit. You will not be able to properly do a force field minimization with a fragment. Better to add a dummy atom or something if that is what you are after. Or create the 3D structure with hydrogens and then replace the proper hydrogens. Since you seem to be working in python, there should be no need to go via the xyz file, you have all the coordinates on the molecule object so use that.
I would do something like:
1. Create molecule from SMILES, possibly with attachment points as deuteriums or similar.
2. Create 3D structure, probably a simple --gen3D would suffice since you are going to add substituents that would change the 3D-structure anyway.
3. Add your attachments
4. Optionally, minimize the resulting structure

This is a rather common procedure, so you should be able to find plenty of example codes by googling...
